ncm-spma/src/main/perl/spma/yumng.pm

```diff
@@ -0,0 +1,810 @@
+    my $wanted_pkgs_a  = Set::Scalar->new(@pkl_a);
+    while ( defined( my $p = $wanted_pkgs_uv->each ) ) {
+        while ( defined( my $p2 = $wanted_pkgs_v->each ) ) {
+            if ( index( $p2, $p ) == 0 ) {
+                my $pkg = $p;
+                chop($pkg);
+                my $pkg_locked = $p2;
+                $pkg_locked =~ tr/;/-/;
+                $self->info("preferring version-locked $pkg_locked over unlocked $pkg");
+                $wanted_pkgs_uv->delete($p);
+            }
+        }
+    }
+    while ( defined( my $p = $wanted_pkgs_uv->each ) ) {
+        chop($p);
+        $wanted_pkgs->insert($p);
+    }
+    while ( defined( my $p = $wanted_pkgs_v->each ) ) {
+        $p =~ s/;/-/;
+        $wanted_pkgs_locked->insert($p);
+    }
+    while ( defined( my $p = $wanted_pkgs_a->each ) ) {
+        $p =~ s/;/./;
+        $wanted_pkgs->insert($p);
+    }
+
+    if ( !$NoAction ) {
+        # Import GPG keys
+        foreach my $file ( glob "/etc/pki/rpm-gpg/RPM-GPG-KEY*" ) {
+            my $gpg = $self->execute_yum_command( ["rpm -v --import $file"], "importing GPG key $file" );
+            return 0 if !defined($gpg);
+        }
+    }
+
+    # Get list of packages installed on system before any package modifications.
+    my $preinstalled_rpms = $self->execute_yum_command( [RPM_QUERY_INSTALLED], "getting list of installed packages", 1 );
+    return 0 if !defined($preinstalled_rpms);
+    $preinstalled_rpms =~ s/\(none\)/0/g;
+    my $preinstalled = Set::Scalar->new( split ( /\n/, $preinstalled_rpms ) );
+
+    if ( !$NoAction && $spma_run ) {
+	# Attempt to update YUM and plugins to latest available release.
+	my $updateyum = $self->execute_yum_command( ["yum update -y yum\* " . YUM_PLUGIN_OPTS], "updating YUM stack" );
+	if ($yum_warned) {
+	    # Be sure to use current version of YUM available in metadata (should handle downgrades).
+	    my $yum_reinstall = $self->execute_yum_command( ["yum reinstall -y yum\* " . YUM_PLUGIN_OPTS], "making sure we are using correct YUM" );
+	    # Expiration of cache is not enough sometimes.
+	    # https://bugzilla.redhat.com/show_bug.cgi?id=1151074
+	    my $yum_clean = $self->execute_yum_command( ["yum clean all " . YUM_PLUGIN_OPTS], "resetting hard YUM state" );
+	    return 0 if !defined($yum_clean);
+	}
+    }
+
+    # Clean up YUM state - worth to be thorough there
+    my $yum_clean = $self->execute_yum_command( ["yum clean expire-cache " . YUM_PLUGIN_OPTS], "expiring YUM caches", 1 );
+    if ($yum_warned) {
+        # Expiration of cache is not enough sometimes.
+        # https://bugzilla.redhat.com/show_bug.cgi?id=1151074
+        my $yum_clean = $self->execute_yum_command( ["yum clean all " . YUM_PLUGIN_OPTS], "resetting hard YUM state", 1 );
+        return 0 if !defined($yum_clean);
+    }
+
+    return 0 if !defined($yum_clean);
+    $yum_clean = $self->execute_yum_command( ["yum makecache " . YUM_PLUGIN_OPTS], "generating YUM cache", 1 );
```

Comment:
The reason for this is to build the initial cache so that subsequent yum/repoquery calls are fast and cached.